MILESTONES.md

```diff
@@ -0,0 +1,152 @@
+          changes come in quickly, but hopefully it's close enough
+
+### Problems:
+
+1. Cursorless commands need to be blocking, and when you use VS Code, they are
+   because of the command server (the action does not return until the command
+   is fully applied.).
+    - However, with the sidecar, the _transmission_ of that cursorless change
+      back to JetBrains would be asynchronous.
+    - This means that something like a cursorless command followed by "air"
+      would have those two commands race.
+        - So you won't really be able to chain cursorless commands alongside
+          other Talon actions unless they do not logically affect the same part
+          of the code.
+        - This is fine for M0.1 because the goal is just to get something
+          working, but will to revisit this for M0.2.
+
+2. The VS Code command server requires a keystroke to be triggered for it to "
+   pick up" the command and respond to it. This requires that VS Code be
+   focused, which obviously wouldn't be the case if the user intends it to be a
+   background sidecar.
+    - On macOS we can work around this because there's an API to send keystrokes
+      to non focused applications, but not on other platforms.
+    - Solution: if you don't care about using VS Code outside of cursorless
+      everywhere / are less paranoid about security, we could fork the command
+      server to be an HTTP server which would not require this keystroke, just
+      for this prototype.
+        - Why wasn't it an HTTP server originally? Security concerns from a
+          user. However it shouldn't be a huge deal -- the existing JetBrains
+          plugin uses a HTTP server with a nonce file.
+        - The other reason we have the keystroke is to provide a synchronization
+          barrier if the user is using VS Code as their primary editor.
+            - Many Talon commands insert by mimicking keystrokes, so the
+              shortcut helps ensure that cursorless commands, which would
+              otherwise run via the asynchronous command server, don't race with
+              the keyboard.
+            - But this isn't important for the Cursorless everywhere user, for
+              whom VS Code only exists to run in the background and handle
+              cursorless commands.
+        - That should unblock @rntz for now.
+
+3. Right now hats in VS Code ("editor decorations") effectively get rebased
+   whenever the user makes changes to the document -- VS Code does that for
+   free.
+    - For example, if you press enter on the line above a series of hats, the
+      hats move down alongside the line instantly -- there's no jank.
+    - This doesn't happen with the JetBrains prototype, so you instead have hats
+      floating in midair
+
+### Other TODOs:
+
+- [ ] Make the sidecar close other tabs when switching files, right now they
+  accumulate
+- [ ] Change the viewport logic so that we center around the cursor, rather than
```

Comment:
Note that when we switch to node.js / in-talon version, this issue will disappear because we can just set things up in our virtual editor so that the virtual size matches JetBrains

Reply:
One question about that - does the current version support noncontiguous ranges of visible lines in a document? For example, in JetBrains you can "collapse" functions or sections of imports single line.

Reply:
Yeah in VSCode the editor object has an attribute called `visibleRanges`, which is an array. If there are no folds, it has one element. With one fold it has two (above and below fold), etc